### PR TITLE
Nixprof: some enchancements

### PR DESCRIPTION
**Pull request by randomstuff**
* Make it available on X11
* Make it work with ELF64
* Use .symtab if available instead of .dynsym
* Only process STT_FUNC symbols
* Use section types and links insteaf of section names to find symbol and string table sections

Some modifications to deps/libelf. Is this code maintained at https://github.com/seL4/libelf? Needs to be upstreamed.

Might be a good idea to use perf or directly perf_event_open (faster, less intrusive, lots of features). For the former, a /tmp/perf-$PID.map can be used to bind SH4 blocks to host addresses.

### CODE REVIEW

**Comment by Holzhaus** [comment]
Looks reasonable to me, although I don't really know what that code does, so @skmp's opinion will be more valueable here.

**Comment by skmp** [comment]
There's no race condition (only single writer), but the design isn't the best one. I'm not sure nixprof still works alright though. There's a C#/windows front end that can visualize the data in a pretty nice way. I wanted to rewrite it to html5 + websockets. Another thing on the wishlist....

**Comment by randomstuff** [comment]
After the fixes, it seems to be working OK.

**Comment by Holzhaus** [comment]
@randomstuff can you rebase? I guess commit 4890e59 has a merge conflict with def93ca.

**Comment by randomstuff** [comment]
Rebased the code.

**Comment by randomstuff** [comment]
I'm probably going to try to implement `/tmp/perf-$pid.map` support which would allow to do basically the same thing with `perf` or the [poor man's profiler](http://poormansprofiler.org/) + some tooling.

**Comment by skmp** [comment]
nixprof was (is?) unique in that it stamped most of the samples and it
works on android (without messing with the kernels). I don't know the state
of things now, though.

On Fri, Aug 28, 2015 at 8:59 PM, Gabriel Corona <notifications@github.com>
wrote:

> I'm probably going to try to implement /tmp/perf-$pid.map support which
> would allow to do basically the same thing with perf or the poor man's
> profiler <http://poormansprofiler.org/> + some tooling.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/reicast/reicast-emulator/pull/807#issuecomment-135860706>
> .
>



-- 
~skmp

**Comment by randomstuff** [comment]
Yes, I don't known how easy it is to make external profiling tools on Android. Apparently, the kernel already has support for perf event (?). Perf timestamps the samples as well but it cannot easily sample [off-cpu](http://www.brendangregg.com/blog/2015-02-26/linux-perf-off-cpu-flame-graph.html) along with on-cpu time.

**Comment by LoungeKatt** [comment]
They're not reliable.

File: core/linux-dist/x11.cpp [hunk 1]
```diff
@@ -72,6 +76,16 @@ void input_x11_handle()
 					{
 						die("death by escape key");
 					}
+#if FEAT_HAS_NIXPROF
+					else if (e.type == KeyRelease && e.xkey.keycode == 76) // F10 button
+					{
+						if (sample_Switch(97)) {
```
Comment (skmp) [inline core/linux-dist/x11.cpp]: 97 hz is wayy too low to get accurate profiling results. I originally used 1K to 10K

**Comment by skmp** [comment]
Well, seeing this breaks nothing, and the changes are minor and in-style, i'm :+1: Check the hz comment. I think I use 3K on android.

File: core/linux-dist/x11.cpp (hunk 1 above)
Comment (randomstuff) [inline core/linux-dist/x11.cpp]: OK, I'll change this.

**Comment by randomstuff** [comment]
Changed the sampling frequency and rebased against master.

**Pull request merged**